File: packages/maps-launcher/android/src/main/java/io/capawesome/capacitorjs/plugins/mapslauncher/classes/options/NavigateOptions.java
package io.capawesome.capacitorjs.plugins.mapslauncher.classes.options;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.JSObject;
import com.getcapacitor.PluginCall;
import io.capawesome.capacitorjs.plugins.mapslauncher.classes.CustomExceptions;

public class NavigateOptions {

    @Nullable
    private final String app;

    @NonNull
    private final Destination destination;

    @Nullable
    private final Destination start;

    @Nullable
    private final String travelMode;

    public NavigateOptions(@NonNull PluginCall call) throws Exception {
        JSObject destinationObject = call.getObject("destination");
        if (destinationObject == null) {
            throw CustomExceptions.DESTINATION_MISSING;
        }
        this.destination = new Destination(destinationObject);
        JSObject startObject = call.getObject("start");
        this.start = startObject == null ? null : new Destination(startObject);
        this.app = call.getString("app");
        this.travelMode = call.getString("travelMode");
    }

    @Nullable
    public String getApp() {
        return app;
    }

    @NonNull
    public Destination getDestination() {
        return destination;
    }

    @Nullable
    public Destination getStart() {
        return start;
    }

    @Nullable
    public String getTravelMode() {
        return travelMode;
    }
}

File: packages/maps-launcher/ios/Plugin/Classes/Options/NavigateOptions.swift
import Foundation
import Capacitor

@objc public class NavigateOptions: NSObject {
    let app: String?
    let destination: Destination
    let start: Destination?
    let travelMode: String?

    init(_ call: CAPPluginCall) throws {
        guard let destinationObject = call.getObject("destination") else {
            throw CustomError.destinationMissing
        }
        self.destination = try Destination(destinationObject)
        if let startObject = call.getObject("start") {
            self.start = try Destination(startObject)
        } else {
            self.start = nil
        }
        self.app = call.getString("app")
        self.travelMode = call.getString("travelMode")
    }
}
